### ISSUE TITLE
Consider adding natural and expectation parameterizations to all exponential families

### ISSUE THREAD

**Issue by NeilGirdhar** [scipy/scipy#6037]
For every exponential family of distribution (normal, gamma, exponential, beta, Dirichlet, etc.) there are two special parametrizations: the natural and expectation parametrizations.  These are useful for implementing generalized linear models and maximum likelihood estimation.

I propose that the following three methods be added to all exponential families:

```
    def nat_to_exp(self, q):
        """
        Convert natural to expectation parameters.
        """

    def exp_to_nat(self, p):
        """
        Convert expectation to natural parameters.
        """

    def sufficient_statistics(self, x):
        """
        Return the sufficient statistics of a sample.
        """
```

And the following methods to all frozen exponential families:

```
    def to_exp(self):
        """
        Return expectation parameters.
        """

    def to_nat(self):
        """
        Return natural parameters.
        """
```

To do maximum likelihood estimation given `n` samples `xs`, you do `np.mean(dist.sufficient_statistics(xs))` and you have the expectation parameters of the maximum likelihood distribution.  E.g., the sufficient statistics of the gamma distribution are x and log(x); the expectation parameters are the mean and the mean-log.  Similarly, the sufficient statistics of the normal distribution are x and x^2, so the expectation parameters are the mean and second moment, which can easily be converted to mean and standard deviation.  (Typically converting from the expectation parametrization to another parametrization needs to use a numerical solver.)

To work with a generalized linear model (that is, to assume that you have `n` independent predictions of the exponential family whose information you want to combine), you simply convert their parametrizations to the natural one and add: `y_nat = np.sum(frozen_dist.to_nat() for frozen_dist in frozen_dists)` gives the natural parametrization of the combination.  E.g., the natural parameters of Bernoulli distribution is log-odds, so the combination is the sum of log-odds.  See [here](http://stats.stackexchange.com/questions/145272/how-is-softmax-unit-derived-and-what-is-the-implication/145277#145277) for some intuition and details.

If you wanted to optimize these distribution (or weights that produced them), the gradient of the cross entropy of some observation `x_exp` and the prediction that was produced above `y_nat` with respect to `y_nat` is the difference of expectation parameters: `x_exp - dist.nat_to_exp(y_nat)`.  For example, in logistic regression, that is, for the Bernoulli distribution, `nat_to_exp` is the logistic sigmoid function.

**Comment by josef-pkt** [comment]
My *first* reaction is that this is to specialized to go into the generic distribution setup.
The more general case are in general that the parameterization is often not the most common one, and there is some need for reparameterization.

This might only work well in the one parameter case.
The scipy distributions are location scale families (in the contiuous case), while some distributions are only in the linear exponential family for some fixed parameters (e.g. NegativeBinomial, Tweedie, and other AFAIR)
In some cases the sufficient statistic is at least bivariate.

(In the regression context I find the "natural" parameterization not "obvious". But that won't be a problem for the constant parameter case, AFAIU.)

The usecase itself is important. statsmodels has many of the linear exponential families in separate classes including derivatives and allowing for different link functions.
We started to write some functions to transform parameters from the mean (scale, extra parameter) GLM parameterization to and from the scipy parameterization but not systematically yet.

Some of the scipy distributions use moment estimator in fit in some cases, e.g. if loc is fixed when lower bound is zero. This would be a similar extension to allow this for the exponential family distributions, but also only applies under some restrictions on the parameters.

**Comment by NeilGirdhar** [comment]
@josef-pkt What do you mean the natural parametrization is "not obvious".  It is unique up to linear map.  It sounds like you don't understand my proposal?

The fact of scipy distributions being location scale families is a design flaw in many cases: location is meaningless for Dirichlet, Beta, Bernoulli, etc.; scale is redundant for distributions that have a concentration (e.g., Von-Mises) and meaningless for distributions that have bounded support (e.g., Dirichlet), and when you add the Von-Mises Fisher distribution on the sphere scale is totally meaningless and can only confuse users (all samples are on the unit sphere, so what should scale mean?)

What I am proposing is the systematic way of working with exponential family distributions.  Tucking this stuff away with GLMs obscures the interchangeability of samples and distribution observations.  Samples are summarized by sufficient statistics and distribution observations are equivalently summarized by expectation parameters.

In general the "link function" _is_ the function that converts from natural to expectation parameters.  Tucking that away in GLM routines hides its true association with a distribution and from users that want to just use the distribution. If one should be implemented in terms of the other, it's the GLM that should be implemented using distribution — not the other way around.

There is only one maximum likelihood estimate of an exponential family distribution and it's the one I outlined above.  Using the first k moments is an approximation (and a bad one depending on the distribution).

**Comment by josef-pkt** [comment]
I find the "natural" GLM parameter pretty useless except for some computational simplifications. (*)

Take for example Bernoulli.
The only obvious parameter is the probability `p`. If we want to link p with some explanatory variables in the regression context then we need a function p(x, beta). which in this case specifies the expectation directly. The properties the the mean function p is an empirical question for the appropriate nonlinearity. In the GLM context the p(x, beta) function is the inverse link and there is a large choice of link functions. In econometrics the most popular is normal cdf because it has a nice interpretation as Probit model, i.e. binary cutoff rule for a latent normal random variable.

If p is constant, then we can estimate it directly, and using the "natural" odds-ratio logistic parameter is not "obvious" to me, in spite of the popularity, log odds ratio is mainly popular because Logit is a common model.

(*) and some theoretical simplification that makes some cases easier to interpret.

(Similar story for Poisson: exp is convenient because it's non-negative, but it's just a functional form assumption that links the intensity parameter to some explanatory variables.
The only "obvious" parameter is the intensity or poisson rate itself.)

**Comment by NeilGirdhar** [comment]
@josef-pkt I don't think you understand what the natural parameter is or why it's useful.

The natural parameter of the Bernoulli distribution is log(p/(1-p)).  The GLM that corresponds to a Bernoulli assumption is logistic regression.

The natural parameter of the beta distribution with shape a, b distribution is (a-1, b-1).  It has a corresponding GLM that I assume is not in scipy.

When estimating the probability p of a Bernoulli, the right estimate is the mean because the sufficient statistics of a Bernoulli is x (the outcome).

However, if you were estimating the shape a, b of a Beta distribution, you would not look at the mean of its samples.  You would average for every sample x, the following statistics log(x) and log(1-x).  From these, the ML estimate of a and b can be found.  These are the minimum statistics that need to be collected, hence "sufficient statistics".

Also, distributional assumptions are not just "functional form assumptions".  Please see my link above for a description of what is being assumed when you use one distribution versus another: namely a set of sufficient statistics and a support.  exp is not just "convenient because it's non-negative" as you say, or else why not use log(1+e^x), which is also non-negative?  There are fundamental assumptions buried in these choices that are obscured by thinking of them as just link functions.

**Comment by NeilGirdhar** [comment]
@josef-pkt I think the source of the confusion here is that "natural parameter" and "expectation parameter" are technical terms.  It might help to read an introduction to exponential families.  Here's a [good introduction](http://arxiv.org/pdf/0911.4863):

Nielsen, Frank, and Vincent Garcia. "Statistical exponential families: A digest with flash cards." arXiv preprint arXiv:0911.4863 (2009).

**Comment by josef-pkt** [comment]
Ok I skimmed the paper, I gave up on reading carefully, because I wasn't interested enough.

I'm doing GLM and similar for 3 years now, and a link function is defined separately from the distribution, whether the link is canonical or not.
I was looking almost only at linear, one parameter exponential family LEF (which is implemented in statsmodels), because it has some very nice statistical properties, e.g. mean is orthogonal to variance or other higher moments in Fisher Information matrix.

In the nonregression case with constant parameters, the parameterizations are all equivalent. However, I don't see that the "natural" parameter is practically very relevant, except maybe in the few cases where we need nonlinear optimization and want a parameter space in R. 
(I haven't seen a use for the duality stuff and Bregman divergence that's in the paper.)

However, the main implementation question is how many cases this actually affects. Based on the flash cards in the paper it doesn't look like many. e.g. Why does Raleigh not have a location parameter? Are most of the two parameter distributions still in the exponential family for all parameters or not? E.g. binomial takes n as fixed, negative binomial takes the shape parameter as fixed (in the LEF), Beta is in LEF for precision fixed AFAIR and so on.

Mean, dispersion parameterization for distribution is very useful, e.g. we (statsmodels PR) have the Beta distribution in the mean-dispersion parameterization which means they can be modeled as two functions of the data.

**Comment by NeilGirdhar** [comment]
@josef-pkt You have a lot of questions that would honestly be best served by reading the _exponential family_ section of an introductory statistics textbook like Shao's _Mathematical Statistics_.  I only linked to the paper since it was brief and succinct, especially section 1.5.

I will try to answer a few of your questions.

The Fisher information matrix of a linear exponential family is just a scalar function of the parameter.  It is the second derivative of the log-normalizer with respect to the natural parameter.

I don't know what you mean by "mean is orthogonal to variance in the Fisher information matrix".  The Fisher information matrix is always independent of location parameters.

**The parametrizations are not "all equivalent"**.  As I explained in my comments above:

> 
> To do maximum likelihood estimation given `n` samples `xs`, you do `np.mean(dist.sufficient_statistics(xs))` and you have the expectation parameters of the maximum likelihood distribution.  E.g., the sufficient statistics of the gamma distribution are x and log(x); the expectation parameters are the mean and the mean-log.  Similarly, the sufficient statistics of the normal distribution are x and x^2, so the expectation parameters are the mean and second moment, which can easily be converted to mean and standard deviation.  (Typically converting from the expectation parametrization to another parametrization needs to use a numerical solver.)
> 
> To work with a generalized linear model (that is, to assume that you have `n` independent predictions of the exponential family whose information you want to combine), you simply convert their parametrizations to the natural one and add: `y_nat = np.sum(frozen_dist.to_nat() for frozen_dist in frozen_dists)` gives the natural parametrization of the combination.  E.g., the natural parameters of Bernoulli distribution is log-odds, so the combination is the sum of log-odds.  See [here](http://stats.stackexchange.com/questions/145272/how-is-softmax-unit-derived-and-what-is-the-implication/145277#145277) for some intuition and details.

You asked how many cases does this actually affect.   I assume you're asking how many distribution (families) are exponential families.  There are an infinite number of such distributions, but in scipy, probably about twenty: Bernoulli, multinomial, beta, Dirichlet, Exponential, geometric, Gamma, normal, Von-Mises, Poisson, negative binomial (for fixed r), logarithmic, Rayleigh, chi (for fixed v, including v=2, the Rayleigh distribution), chi-squared, Laplace, Weibull, Wishart, inverse-Wishart.  Notably missing: Von-Mises Fisher.

Beta is a linear (curved) exponential family if you fix a linear transformation of the natural parameters (a-1, b-1), but the precision is not a linear function of that.

I'm not suggesting that other parametrizations are not useful.   What I am saying is that the natural and expectation parametrizations should be exposed.

**Comment by josef-pkt** [comment]
> You asked how many cases does this actually affect. I assume you're asking how many distribution (families) are exponential families.

The question was more specific to which distributions have a "natural" parameterization for all the parameters. The flash cards in the paper have some gaps.
I didn't check now specific cases, but many distribution in the literature have 1, 2, or 3 parameter versions. For which does it apply?

In case I wasn't clear enough. I read tons (literally less than 100kg printed) of books and papers on GLM/LEF. But I don't really care about the constant parameter case. The standard regression case is the mean model `mean = f(X, beta)` where X are explanatory variables (like in Logit/Logistic regression) and the scale/dispersion is a function of the mean and some other parameters.

In this case choosing different `f` or link functions implies different nonlinear specification of the effect of the explanatory variables on the mean of the response variable.

If there are no explanatory variables, or X is just a vector of ones, then there is no effect of the nonlinearity. Which parameterization is chosen is just for convenience, for example to use sufficient statistics directly, they all imply the same model.

(aside: for hypothesis testing like comparing rates and proportions, the transformation of the parameters are often chosen to make them closer to the normal distribution, which is again a different transformation, e.g. atan for ???, sqrt for Poisson, ...)

> What I am saying is that the natural and expectation parametrizations should be exposed.

An additional method is not really that costly, however the distribution classes are pretty generic and encompass around 100 distributions. Is it worth adding it for a few special cases?

Note: Looking systematically at sufficient statistics is a good idea, and I guess the corresponding `_fit` methods might be missing for several distributions.

**Comment by NeilGirdhar** [comment]
> The question was more specific to which distributions have a "natural" parameterization for all the parameters

All exponential families have a natural parametrization.  It is fundamental to being an exponential family.

> Which parameterization is chosen is just for convenience, for example to use sufficient statistics directly, they all imply the same model.

I'm sorry, but no.  Using the natural parametrization for your explanatory variables corresponds to an independence assumption of explanatory distributions, because addition in this space uniquely (up to linear map) corresponds to a pointwise product of densities of the distributions.  Here's why:  If you can write your density as f(x; m) = exp(m•x + g(m) + h(x)) then m are your natural parameters.  the pointwise product of densities when h=0 is exp(m_1•x) * exp(m_2•x) renormalized.  This is just exp((m_1+m_2)•x + g(m_1 + m_2)).  Another parametrization k(m) where k is a nonlinear function cannot have this property.

> An additional method is not really that costly, however the distribution classes are pretty generic and encompass around 100 distributions. Is it worth adding it for a few special cases?

It's 20 special distribution that happen to be some of the most common distributions especially given the popularity of Bayesian inference.   I think these methods should be put into a subclass for exponential families.

> Note: Looking systematically at sufficient statistics is a good idea, and I guess the corresponding _fit methods might be missing for several distributions.

If `_fit` were implemented as I suggested above, it would always fine the maximum likelihood solution.

**Comment by josef-pkt** [comment]
In the constant parameter case there is a one-to-one mapping between all parameterizations, so they all represent the same model/distribution with just different parameterization.
Using one parameterization for a proof or computation doesn't change that.

I was talking about explanatory variables, not aggregation of independent samples.
https://en.wikipedia.org/wiki/Generalized_linear_model
where choosing link function is a separate issue.

There are some special properties of canonical link functions that are exploited in some models.
Also, I'm not yet very clear about the implication for aggregation
quote in this issue https://github.com/statsmodels/statsmodels/issues/2840
AFAICS aggregation is very sensitive to misspecification, e.g. heterogeneity as in #6000

---

In any case, the question is whether it's large enough to provide it as special cases for some distributions.

Related: I was thinking that it would be nice to have hazard function and characteristic function, or moment generating functions, for distributions where those functions have an explicit form. But I never checked how many distributions this would affect. I'm usually just interested in a few.

**Comment by NeilGirdhar** [comment]
> Using one parameterization for a proof or computation doesn't change that.

The point is that** additivity in these parametrizations does make a difference**.  It's not just using something because it's convenient for a proof.

Choosing a link function is not a separate issue.  Each of the link functions in that article corresponds to the gradient-log-normalizer of an exponential family, which is the `nat_to_exp` function I'm proposing.

Exposing the moments is interesting too, but a different issue.

**Comment by josef-pkt** [comment]
I didn't try to figure out what the gradient-log-normalizer is in terms of standard GLM terminology.
AFAIU, your `nat_to_exp` is just the (inverse) canonical link function where "eta" (linear predictor) = "nat"

In terms of the paper, what we would need in GLM is original -> expectation or \Lambda -> h  
which is trivial in Poisson and Binomial. This would allow us to move between the estimated expectation plus auxiliary parameters to the scipy distribution.

Once we have the expectation/mean, then the link function links expectation (of mean) to the explanatory variables, or more precisely to the linear predictor. This is a separate step based on the expectation parameterization. Setting `nat` equal to linear predictor in GLM is just a special case with nice properties, but with functional form restriction.

Going `original -> nat -> exp -> linear predictor/eta` would just be a detour.

In any case, whether I find canonical links ("natural" parameterization) of general interest or of limited usefulness is not the main issue.

**Comment by NeilGirdhar** [comment]
> I didn't try to figure out what the gradient-log-normalizer is in terms of standard GLM terminology.

The gradient-log-normalizer is `nat_to_exp`.  For the Bernoulli distribution, it is the logistic function.   All of the link functions on the GLM page you linked are this conversion from natural to expectation parametrization.

> Going original -> nat -> exp -> linear predictor/eta would just be a detour.

I don't know what you're talking about, but to implement a GLM _all you need_ is `nat_to_exp`.

**Comment by NeilGirdhar** [comment]
Since I filed this issue, I ended up writing my own exponential family library in JAX: https://github.com/NeilGirdhar/efax/
